server/src/main/java/io/druid/server/QueryManager.java

```diff
@@ -53,16 +66,45 @@ public void registerQuery(Query query, final ListenableFuture future)
   {
     final String id = query.getId();
     queries.put(id, future);
+    if(queryDatasources.get(id) == null) {
+      queryDatasources.put(id, getDataSources(query));
+    }
     future.addListener(
         new Runnable()
         {
           @Override
           public void run()
           {
             queries.remove(id, future);
+            queryDatasources.remove(id);
           }
         },
         MoreExecutors.sameThreadExecutor()
     );
   }
+
+  public List<String> getQueryDatasources(final String queryId) {
+    return queryDatasources.get(queryId);
+  }
+
+  public List<String> getDataSources(final Query query) {
+    List<String> datasources = new ArrayList<>();
+    getDataSourcesHelper(query, datasources);
+    return datasources;
+  }
+
+  private void getDataSourcesHelper(final Query query, List<String> datasources) {
+    if (query.getDataSource() instanceof TableDataSource) {
+      // there will only be one datasource for TableDataSource
+      datasources.addAll(query.getDataSource().getNames());
```

Comment:
Looking at the getNames call on the DataSource impls here, it looks like you *should* be able to just use getNames